### PR TITLE
DCS-738 BVL add cancel button to amend bookings

### PR DESCRIPTION
**Pull request by GurnankCheema**
New page created that informs user that the revised appointment requested is not available.
The page displays the date and time of the requested booking. 
It also has a continue button that takes user back to the previous page where they can request a different date/time combination.
Lastly it has a cancel button that takes user to the current appointment details page

Screenshot 2021-01-21 at 17 46 09

### CODE REVIEW

File: views/amendBooking/noAvailabilityForDateTime.njk [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+            </form>
+
+            <p class="govuk-body">
+                <a href="/booking-details/{{bookingId}}" class="govuk-link" data-qa="cancel">Cancel changes</a>
+            </p>
+
+        </div>
+    </div>
+{% endblock %}
```
Comment (andrewrlee) [inline views/amendBooking/noAvailabilityForDateTime.njk:43]: needs a new line

File: integration-tests/plugins/index.js [hunk 2]
```diff
@@ -33,7 +33,7 @@ module.exports = on => {
     stubGroups: caseload => whereabouts.stubGroups(caseload),
     stubCreateVideoLinkBooking: () => whereabouts.stubCreateVideoLinkBooking(),
     getBookingRequest: () => whereabouts.getBookingRequest(),
-    getFindBookingRequest: () => whereabouts.getFindBookingRequest(),
+    getFindBookingRequests: () => whereabouts.getFindBookingRequests(),
```
Comment (andrewrlee) [inline integration-tests/plugins/index.js:36]: could do with renaming this

File: backend/routes/amendBooking/videoLinkNotAvailableController.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        startTimeMinutes: req.body.startTimeMinutes,
+        endTimeHours: req.body.startTimeMinutes,
+        endTimeMinutes: req.body.endTimeMinutes,
+        preAppointmentRequired: req.body.preAppointmentRequired,
+        postAppointmentRequired: req.body.postAppointmentRequired,
+        startTime: req.body.startTime,
+        endTime: req.body.endTime,
+        prisoner: req.body.prisoner,
+        locations: req.body.locations,
+      }
```
Comment (andrewrlee) [inline backend/routes/amendBooking/videoLinkNotAvailableController.ts:28]: read and write out using  new codec

File: backend/routes/amendBooking/changeDateAndTimeController.ts [hunk 4]
```diff
@@ -42,7 +49,75 @@ export = class ChangeDateAndTimeController {
...
+        startTime: startTime.format('HH:mm'),
+        endTime: endTime.format('HH:mm'),
+        prisoner: {
+          name: prisonerName,
+        },
+        locations: {
+          prison: prisonName,
+          court: courtLocation,
+        },
+      }
```
Comment (andrewrlee) [inline backend/routes/amendBooking/changeDateAndTimeController.ts:101]: read and write using codec

File: backend/routes/amendBooking/changeDateAndTimeController.ts [hunk 5]
```diff
@@ -42,7 +49,75 @@ export = class ChangeDateAndTimeController {
...
+
+      return res.redirect(`/video-link-not-available/${bookingId}`)
     }
   }
 }
+
+function parseAvailabilityRequest(agencyId: string, obj: Record<string, unknown>): AvailabilityRequest {
+  return {
+    agencyId,
```
Comment (andrewrlee) [inline backend/routes/amendBooking/changeDateAndTimeController.ts:116]: parse into form

File: backend/routes/amendBooking/changeDateAndTimeController.test.ts [hunk 6]
```diff
@@ -237,13 +249,35 @@ describe('change date and time controller', () => {
...
+      }
+      availabilityCheckService.getAvailability.mockResolvedValue(availability)
       req.params.bookingId = '12'
 
       await controller.submit(false)(req, res, null)
 
       expect(res.redirect).toHaveBeenCalledWith(`/video-link-available/12`)
     })
 
+    it("should display the '/video-link-not-available' page on submit", async () => {
```
Comment (andrewrlee) [inline backend/routes/amendBooking/changeDateAndTimeController.test.ts:266]: "should redirect when not available"

**Review by andrewrlee — APPROVED** [review]
LGTM

File: views/amendBooking/noAvailabilityForDateTime.njk (hunk 1 above)
Comment (GurnankCheema) [reply views/amendBooking/noAvailabilityForDateTime.njk:43]: done

File: backend/routes/amendBooking/changeDateAndTimeController.test.ts (hunk 6 above)
Comment (GurnankCheema) [reply backend/routes/amendBooking/changeDateAndTimeController.test.ts:266]: done. 
Although I do feel it is better to be more specific when describing what you expect to happen.

**Pull request merged**